tket/tests/test_GlobalisePhasedX.cpp

```diff
@@ -0,0 +1,330 @@
+// Copyright 2019-2021 Cambridge Quantum Computing
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <catch2/catch.hpp>
+
+#include "Circuit/Circuit.hpp"
+#include "Simulation/CircuitSimulator.hpp"
+#include "Transformations/Decomposition.hpp"
+#include "testutil.hpp"
+
+namespace tket {
+namespace test_GlobalisePhasedX {
+
+static bool is_global(Vertex v, const Circuit& c) {
+  unsigned in = c.n_in_edges_of_type(v, EdgeType::Quantum);
+  unsigned out = c.n_out_edges_of_type(v, EdgeType::Quantum);
+  return in == c.n_qubits() && out == c.n_qubits();
+}
+
+SCENARIO("globalise_PhasedX transform") {
+  GIVEN("A simple circuit") {
+    Circuit c1(2);
+    c1.add_op<unsigned>(OpType::NPhasedX, {0.4, 0.3}, {0, 1});
+    c1.add_op<unsigned>(OpType::NPhasedX, {0.4, 0.3}, {0, 1});
+
+    WHEN("Applying the transform") {
+      Circuit c2 = c1;
+      REQUIRE(Transforms::globalise_PhasedX().apply(c2));
+      THEN("There are two global NPhasedX gates") {
+        REQUIRE(c2.count_gates(OpType::NPhasedX) == 2);
+        BGL_FORALL_VERTICES(v, c2.dag, DAG) {
+          OpType type = c2.get_OpType_from_Vertex(v);
+          if (type == OpType::NPhasedX) {
+            REQUIRE(is_global(v, c2));
+          } else if (c2.detect_singleq_unitary_op(v)) {
+            // single-qb gates are Rz
+            REQUIRE(type == OpType::Rz);
+          }
+        }
+      }
+      THEN("The resulting unitaries are equal") {
+        auto u1 = tket_sim::get_unitary(c1);
+        auto u2 = tket_sim::get_unitary(c2);
+        REQUIRE((u1 - u2).cwiseAbs().sum() < ERR_EPS);
```

Comment:
Instead of `cwiseAbs().sum()` can we use eigen's `isApprox()` (here and elsewhere)?

Reply:
Done.